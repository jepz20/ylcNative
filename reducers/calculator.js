// @flow

import { Record } from 'immutable'
import type { RecordFactory, RecordOf } from 'immutable'
import {
  DIGIT_PRESS,
  ERASE,
  SET_CURRENT_PLAYER,
  TOGGLE_CALCULATOR_VISIBILITY,
  RESET_CALCULATOR_VALUE
} from '../actions/types'
import type { Action } from '../actions/calculator'
import {
  COMMON_VALUE_LENGTH,
  MAX_VALUE_LENGTH,
  CALCULATOR_DEFAULT_VALUE
} from '../constants'

import { parsePoints } from '../utils'
export type State = {
  value: string,
  currentPlayer: string | null,
  visible: boolean
}

export type StateRE = RecordOf<State>

export const INITIAL_STATE: RecordFactory<State> = Record({
  value: CALCULATOR_DEFAULT_VALUE,
  currentPlayer: null,
  visible: false
})

const getCalculatedValue = (currentValue: string, digit: string): string => {
  if (currentValue === CALCULATOR_DEFAULT_VALUE) {
    if (parsePoints(digit) === parsePoints(CALCULATOR_DEFAULT_VALUE)) {
      return CALCULATOR_DEFAULT_VALUE
    }
    return digit
  }

  if (currentValue.length >= MAX_VALUE_LENGTH) {
    return currentValue
  }

  let newValue = `${currentValue}${digit}`

  // if the value contains more than one digit(ej: 00, 000) and the length
  // is too big the user probably pressed it by mistake in that case just put
  // one 0
  let diff = newValue.length - COMMON_VALUE_LENGTH
  if (diff > 0 && digit.length > 1) {
    // put as much digits as necessary to have the common value length
    newValue = `${currentValue}${digit.slice(0, -diff) || 0}`
  }

  if (newValue.length > MAX_VALUE_LENGTH) {
    newValue = newValue.slice(0, MAX_VALUE_LENGTH)
  }

  return newValue
}

const getErasedValue = (currentValue: string): string => {
  return currentValue.slice(0, -1) || CALCULATOR_DEFAULT_VALUE
}

export default function (
  state: StateRE = INITIAL_STATE(),
  action: Action
): StateRE {
  const { value } = state
  switch (action.type) {
    case DIGIT_PRESS:
      return state.merge({
        value: getCalculatedValue(value, action.payload.digit)
      })
    case ERASE:
      return state.merge({ value: getErasedValue(value) })
    case TOGGLE_CALCULATOR_VISIBILITY:
      return state.merge({
        visible: !state.visible,
        currentPlayer: action.payload.player
      })
    case SET_CURRENT_PLAYER: {
      return state.merge({ currentPlayer: action.payload.player })
    }
    case RESET_CALCULATOR_VALUE: {
      return state.merge({ value: CALCULATOR_DEFAULT_VALUE })
    }
    default:
      return state
  }
}
