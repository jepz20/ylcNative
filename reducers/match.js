// @flow

import { Record, Map, List } from 'immutable'
import type { RecordFactory } from 'immutable'
import {
  SCOOP_MATCH,
  DRAW,
  HALF_POINTS,
  ADD_POINTS,
  SUBSTRACT_POINTS
} from '../actions/types'
import type {
  StateRE,
  State,
  Result,
  Player,
  Log,
  LogList,
  LogMap,
  PlayerMap,
  ResultMap
} from '../types/match'
import type { Action } from '../actions/match'
import { parsePoints } from '../utils'

// TODO: Use AsyncStorage instead of hardcoded values
const DEFAULT_POINTS: number = 8000
const PLAYER_1_NAME: string = 'Jose'
const PLAYER_2_NAME: string = 'Eduardo'

const getId = (): string => Date.now().toString()

export const PlayerRecord: RecordFactory<Player> = Record({
  id: getId(),
  name: 'Default',
  currentPoints: DEFAULT_POINTS
})

export const LogsRecord: RecordFactory<Log> = Record({
  id: getId(),
  playerId: '0',
  operationValue: 0,
  currentPoints: 0,
  previousPoints: 0
})

export const resultsRecord: RecordFactory<Result> = Record({
  winner: null
})

const INITIAL_STATE: RecordFactory<State> = Record({
  players: Map({
    '1': PlayerRecord({
      id: '1',
      name: PLAYER_1_NAME,
      currentPoints: DEFAULT_POINTS
    }),
    '2': PlayerRecord({
      id: '2',
      name: PLAYER_2_NAME,
      currentPoints: DEFAULT_POINTS
    })
  }),
  currentDuel: '1',
  logs: Map(),
  results: Map()
})

const getPlayerPoints = (players: PlayerMap, player: string): number => {
  if (players.has(player)) {
    // $FlowFixMe
    return players.get(player).currentPoints
  }
  console.warn(players, player, 'getPlayerPoint')
  return 0
}

const addToLog = (
  currentDuel: string,
  logs: ?LogMap,
  playerId: string,
  operationValue: number,
  previousPoints: number,
  nextPoints: number
): LogMap => {
  const newLogs: LogMap = logs || Map()
  const newLog: LogList = (newLogs.get(currentDuel) || List()).push(
    LogsRecord({
      id: getId(),
      playerId,
      operationValue,
      currentPoints: nextPoints,
      previousPoints
    })
  )
  return newLogs.merge({ [currentDuel]: newLog })
}

const setDuelResult = (
  players: PlayerMap,
  currentDuel: string,
  results: ?ResultMap = Map()
): ResultMap => {
  const activePlayers: PlayerMap = players.filter(
    player => player.currentPoints > 0
  )

  const res: ResultMap = results || Map()
  const isWinner: boolean = activePlayers.size === 1
  const isTie: boolean = !activePlayers.size

  if (isWinner) {
    return res.merge({
      [currentDuel]: resultsRecord({
        // $FlowFixMe
        winner: activePlayers.first().id
      })
    })
  }

  if (isTie) {
    return res.merge({
      [currentDuel]: resultsRecord({ winner: 'tie' })
    })
  }

  return res
}

const operateValue = (
  state: StateRE,
  player: string,
  operationValue: number
): StateRE => {
  const previousPoints: number = getPlayerPoints(state.players, player)
  const nextPoints: number = previousPoints + operationValue

  const players = state.players.mergeDeep({
    [player]: {
      currentPoints: nextPoints
    }
  })
  const { currentDuel, logs, results } = state
  return state.merge({
    players: players,
    logs: addToLog(
      currentDuel,
      logs,
      player,
      operationValue,
      previousPoints,
      nextPoints
    ),
    results: setDuelResult(players, currentDuel, results)
  })
}

const draw = (state: StateRE): StateRE => {
  const players: PlayerMap = state.players.map(player =>
    player.set('currentPoints', 0)
  )
  return state.merge({
    players,
    results: setDuelResult(players, state.currentDuel, state.results)
  })
}

export default function (state: StateRE = INITIAL_STATE(), action: Action) {
  switch (action.type) {
    case ADD_POINTS: {
      const { points, player } = action.payload
      let operationValue: number = parsePoints(points)
      return operateValue(state, player, operationValue)
    }
    case SUBSTRACT_POINTS: {
      const { points, player } = action.payload
      let operationValue: number = parsePoints(points) * -1
      return operateValue(state, player, operationValue)
    }
    case HALF_POINTS: {
      const { player } = action.payload
      let operationValue: number =
        Math.ceil(getPlayerPoints(state.players, player) / 2) * -1
      return operateValue(state, player, operationValue)
    }
    case SCOOP_MATCH: {
      const { player } = action.payload
      let operationValue: number =
        Math.ceil(getPlayerPoints(state.players, player)) * -1
      return operateValue(state, player, operationValue)
    }
    case DRAW:
      return draw(state)
    default:
      return state
  }
}
