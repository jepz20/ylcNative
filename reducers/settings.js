const INITIAL_STATE = {
  duelsPerMatch: 3,
  theme: 'defaul'
}

export default function (state = INITIAL_STATE, action) {
  switch (action.type) {
    default:
      return state
  }
}
