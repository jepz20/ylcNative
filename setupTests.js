// setup file
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import configureStore from 'redux-mock-store'
import PropTypes from 'prop-types'
import React from 'react'

configure({ adapter: new Adapter() })

export class TestProvider extends React.Component {
  getChildContext () {
    return { store: this.store }
  }

  constructor (props, context) {
    super(props)
    this.store = props.store
  }

  render () {
    return this.props.children
  }
}

TestProvider.childContextTypes = {
  store: PropTypes.object
}

export const mockStore = configureStore()
