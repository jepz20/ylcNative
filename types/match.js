// @flow

import type { RecordOf, Map, List } from 'immutable'
export type Player = {
  id: string,
  name: string,
  currentPoints: number
}

export type Log = {
  id: string,
  playerId: string,
  operationValue: number,
  currentPoints: number,
  previousPoints: number,
  type: string
}

export type Result = {
  winner: ?string
}

export type PlayerRE = RecordOf<Player>
export type LogRE = RecordOf<Log>
export type ResultRE = RecordOf<Result>
export type LogList = List<LogRE>

export type PlayerMap = Map<string, PlayerRE>
export type LogMap = Map<string, LogList>
export type ResultMap = Map<string, ResultRE>

export type State = {
  players: PlayerMap,
  currentDuel: string,
  logs: LogMap,
  results: ResultMap
}

export type StateRE = RecordOf<State>
