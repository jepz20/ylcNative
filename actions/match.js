// @flow

import * as T from './types'

type PlayerPayload = {
  player: string,
  logId: string
}

type PointsPayload = {
  points: string
}

type ScoopMatchAction = {
  type: T.SCOOP_MATCH_STR,
  payload: PlayerPayload
}

type DrawAction = {
  type: T.DRAW_STR,
  payload: { logId: string }
}

type HalfPointsAction = {
  type: T.HALF_POINTS_STR,
  payload: PlayerPayload
}

type AddPointsAction = {
  type: T.ADD_POINTS_STR,
  payload: PlayerPayload & PointsPayload
}

type SubstractPointsAction = {
  type: T.SUBSTRACT_POINTS_STR,
  payload: PlayerPayload & PointsPayload
}

type ChangePlayerNameAction = {
  type: T.CHANGE_PLAYER_NAME_STR,
  payload: {
    name: string,
    player: string
  }
}

export type Action =
  | ScoopMatchAction
  | DrawAction
  | HalfPointsAction
  | AddPointsAction
  | SubstractPointsAction
  | ChangePlayerNameAction

export type ScoopMatch = (player: string, logId: string) => ScoopMatchAction
export const scoopMatch: ScoopMatch = (player, logId) => ({
  type: T.SCOOP_MATCH,
  payload: { player, logId }
})

export type Draw = (logId: string) => DrawAction
export const draw: Draw = logId => ({
  type: T.DRAW,
  payload: {
    logId
  }
})

export type HalfPoints = (player: string, logId: string) => HalfPointsAction
export const halfPoints: HalfPoints = (player, logId) => ({
  type: T.HALF_POINTS,
  payload: { player, logId }
})

export type AddPoints = (
  points: string,
  player: string,
  logId: string
) => AddPointsAction
export const addPoints: AddPoints = (points, player, logId) => ({
  type: T.ADD_POINTS,
  payload: { points, player, logId }
})

export type SubstractPoints = (
  points: string,
  player: string,
  logId: string
) => SubstractPointsAction
export const substractPoints: SubstractPoints = (points, player, logId) => ({
  type: T.SUBSTRACT_POINTS,
  payload: { points, player, logId }
})

export type ChangePlayerName = (
  name: string,
  player: string
) => ChangePlayerNameAction
export const changePlayerName: ChangePlayerName = (name, player) => ({
  type: T.CHANGE_PLAYER_NAME,
  payload: { name, player }
})
