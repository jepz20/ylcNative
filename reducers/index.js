import { combineReducers } from 'redux'
import match from './match'
import settings from './settings'
import calculator from './calculator'

export default combineReducers({
  match,
  settings,
  calculator
})
