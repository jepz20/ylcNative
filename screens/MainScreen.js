import React, { Component } from 'react'
import { View, StyleSheet } from 'react-native'
import { connect } from 'react-redux'
import * as actions from '../actions'
import { DuelHistoryContainer, LifePointsContainer } from '../containers'

class Main extends Component {
  toggleCalculatorVisibility = (player) => {
    const { toggleCalculatorVisibility, setCurrentPlayer } = this.props
    setCurrentPlayer(player)
    toggleCalculatorVisibility()
  }

  render () {
    const {
      match: { logs, players, matchResults, currentDuel }, settings } = this.props
    return (
      <View style={styles.container}>
        <View style={styles.LifePointsContainerWrapper}>
          <LifePointsContainer
            players={players}
            matchResults={matchResults}
            settings={settings}
            onPress={this.toggleCalculatorVisibility}
          />
        </View>
        <DuelHistoryContainer
          style={styles.DuelHistoryContainer}
          logs={logs}
          currentDuel={currentDuel}
          players={players}
          onPress={this.toggleCalculatorVisibility}
        />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 60
  },
  LifePointsContainerWrapper: {
    flexDirection: 'row',
    flex: 2,
    backgroundColor: 'green'
  },
  DuelHistoryContainer: {
    flex: 5
  }
})

const mapStateToProps = ({ match, settings }) => ({
  match,
  settings
})

export default connect(mapStateToProps, actions)(Main)
