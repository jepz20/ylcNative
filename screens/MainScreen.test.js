import React from 'react'
import renderer from 'react-test-renderer'
import { TestProvider, mockStore } from '../setupTests'
import MainScreen from './MainScreen'
import { INITIAL_STATE } from '../reducers/match'

jest.mock('LayoutAnimation', () => {
  // console.log('LAYOUTANIMATION')
  return {
    easeInEaseOut: jest.fn()
  }
})

describe('MainScreen', () => {
  test('renders correctly', () => {
    const store = mockStore({ match: INITIAL_STATE() })
    const mainScreen = renderer
      .create(
        <TestProvider store={store}>
          <MainScreen />
        </TestProvider>
      )
      .toJSON()
    expect(mainScreen).toMatchSnapshot()
  })
})
