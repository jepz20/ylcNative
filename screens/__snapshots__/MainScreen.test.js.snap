// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 60,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "green",
        "flex": 2,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "green",
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
        testID={undefined}
      >
        <View
          style={
            Object {
              "backgroundColor": "#006400",
              "borderColor": "#005a00",
              "borderWidth": 1,
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "height": 35,
                "justifyContent": "space-around",
                "marginTop": 10,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontSize": 65,
                    "fontWeight": "bold",
                  },
                  Object {
                    "fontSize": 65,
                  },
                ]
              }
            >
              8000
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 40,
                "justifyContent": "flex-end",
                "marginBottom": 5,
                "marginLeft": 5,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              onChangeText={[Function]}
              style={
                Object {
                  "borderBottomWidth": 0,
                  "color": "#fff",
                  "fontSize": 17,
                  "fontWeight": "bold",
                }
              }
              underlineColorAndroid="transparent"
              value="Jose"
            />
          </View>
        </View>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
        testID={undefined}
      >
        <View
          style={
            Object {
              "backgroundColor": "#006400",
              "borderColor": "#005a00",
              "borderWidth": 1,
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "height": 35,
                "justifyContent": "space-around",
                "marginTop": 10,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontSize": 65,
                    "fontWeight": "bold",
                  },
                  Object {
                    "fontSize": 65,
                  },
                ]
              }
            >
              8000
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 40,
                "justifyContent": "flex-end",
                "marginBottom": 5,
                "marginLeft": 5,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              onChangeText={[Function]}
              style={
                Object {
                  "borderBottomWidth": 0,
                  "color": "#fff",
                  "fontSize": 17,
                  "fontWeight": "bold",
                }
              }
              underlineColorAndroid="transparent"
              value="Eduardo"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flex": 5,
          "flexDirection": "row",
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flex": 1,
          "justifyContent": "flex-start",
        }
      }
    >
      <View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={undefined}
          testID={undefined}
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 10,
                "marginTop": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "grey",
                  "fontSize": 30,
                }
              }
            >
              .
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 10,
                "marginTop": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "grey",
                  "fontSize": 30,
                }
              }
            >
              .
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 10,
                "marginTop": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "grey",
                  "fontSize": 30,
                }
              }
            >
              .
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <RCTScrollView
      contentContainerStyle={
        Object {
          "flex": 1,
          "justifyContent": "flex-start",
        }
      }
    >
      <View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={undefined}
          testID={undefined}
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 10,
                "marginTop": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "grey",
                  "fontSize": 30,
                }
              }
            >
              .
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 10,
                "marginTop": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "grey",
                  "fontSize": 30,
                }
              }
            >
              .
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 10,
                "marginTop": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "alignSelf": "flex-end",
                  "color": "grey",
                  "fontSize": 30,
                }
              }
            >
              .
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
